controlpanel/api/cluster.py

```diff
@@ -463,121 +463,139 @@ class ToolDeploymentError(Exception):
-            # f'--version', tool.version,
-            f'--namespace', user.k8s_namespace,
-            *set_values,
-        )
+    @property
+    def chart_name(self):
+        return self.tool.chart_name
 
-    except helm.HelmError as error:
-        raise ToolDeploymentError(error)
-
-
-def list_tool_deployments(user, id_token, search_name=None, search_version=None):
-    deployments = []
-    k8s = KubernetesClient(id_token=id_token)
-    results = k8s.AppsV1Api.list_namespaced_deployment(user.k8s_namespace)
-    for deployment in results.items:
-        app_name = deployment.metadata.labels["app"]
-        _, version = deployment.metadata.labels["chart"].rsplit("-", 1)
-        if search_name and search_name not in app_name:
-            continue
-        if search_version and not version.startswith(search_version):
-            continue
-        deployments.append(deployment)
-    return deployments
-
-
-def get_tool_deployment(tool_deployment, id_token):
-    deployments = list_tool_deployments(
-        tool_deployment.user,
-        id_token,
-        search_name=tool_deployment.tool.chart_name,
-        # search_version=tool_deployment.tool.version,
-    )
+    @property
+    def k8s_namespace(self):
+        return self.user.k8s_namespace
 
-    if not deployments:
-        raise ObjectDoesNotExist(tool_deployment)
+    @property
+    def release_name(self):
+        return f"{self.chart_name}-{self.user.slug}"
+
+    def install(self, **kwargs):
+        values = {
+            "username": self.user.username.lower(),
+            "Username": self.user.username.lower(),  # XXX backwards compatibility
+            "aws.iamRole": self.user.iam_role_name,
+            "toolsDomain": settings.TOOLS_DOMAIN,
+        }
 
-    if len(deployments) > 1:
-        log.warning(f"Multiple matches for {tool_deployment!r} found")
-        raise MultipleObjectsReturned(tool_deployment)
+        # generate per-deployment secrets
+        for key, value in self.tool.values.items():
+            if value == "<SECRET_TOKEN>":
+                self.tool.values[key] = secrets.token_hex(32)
 
-    return deployments[0]
+        values.update(self.tool.values)
+        values.update(kwargs)
+        set_values = []
+        for key, val in values.items():
+            escaped_val = val.replace(',', '\,')
+            set_values.extend(['--set', f'{key}={escaped_val}'])
 
+        try:
+            return helm.upgrade_release(
+                self.release_name,
+                f'{settings.HELM_REPO}/{self.chart_name}',  # XXX assumes repo name
+                # f'--version', tool.version,
+                f'--namespace', self.k8s_namespace,
+                *set_values,
+            )
+
+        except HelmError as error:
+            raise ToolDeploymentError(error)
+
+    def uninstall(self, id_token):
+        deployment = self.get_deployment(id_token)
+        helm.delete(
+            deployment.metadata.name,
```

Comment:
@andyhd do you remember why when we install the tool charts we use ` f"{self.chart_name}-{self.user.slug}"` as release name (now moved in `ToolDeployment.release_name` property) but when we uninstall we get the `Deployment` name from k8s?

Not a big deal, I've left as it is for now (as this is just a refactoring) but potentially we could remove the need to make a request to the k8s API and just run the `helm delete`.

Reply:
(plus if we don't talk with k8s API we don't need the `id_token` parameter)

Reply:
I think I was planning to use Kubernetes generated release names instead of specifying them. But yes, reuse `release_name` here and avoid the API call. 👍

Reply:
That would be nice, would that break with some of the installed releases you reckon?

Reply:
To be fair, we currently don't expose the "uninstall" tool functionality, still left for next round of refactorings as don't like to change too much at the same time ☺️ 

(I'm testing these changes in `dev` again and it looks like I didn't break anything so release to `alpha` will probably follow soon)